use-cases/elastic_router/ddtcpreplay/src/tcpreplay.1
.de1 NOP
.  it 1 an-trap
.  if \\n[.$] \,\\$*\/
..
.ie t \
.ds B-Font [CB]
.ds I-Font [CI]
.ds R-Font [CR]
.el \
.ds B-Font B
.ds I-Font I
.ds R-Font R
.TH tcpreplay 1 "11 Jan 2016" "tcpreplay" "User Commands"
.\"
.\" DO NOT EDIT THIS FILE (in-mem file)
.\"
.\" It has been AutoGen-ed
.\" From the definitions tcpreplay_opts.def
.\" and the template file agman-cmd.tpl
.SH NAME
\f\*[B-Font]tcpreplay\fP
\- Replay network traffic stored in pcap files
.SH SYNOPSIS
\f\*[B-Font]tcpreplay\fP
.\" Mixture of short (flag) options and long options
[\f\*[B-Font]\-flags\f[]]
[\f\*[B-Font]\-flag\f[] [\f\*[I-Font]value\f[]]]
[\f\*[B-Font]\-\-option-name\f[][[=| ]\f\*[I-Font]value\f[]]]
<pcap_file(s)>
.sp \n(Ppu
.ne 2

tcpreplay is a tool for replaying network traffic from files saved with
tcpdump or other tools which write pcap(3) files.
.SH "DESCRIPTION"
The basic operation of tcpreplay is to resend  all  packets  from  the
input file(s) at the speed at which they were recorded, or a specified 
data rate, up to as fast as the hardware is capable.
.sp
Optionally, the traffic can be split between two interfaces, written to
files, filtered and edited in various ways, providing the means to test
firewalls, NIDS and other network devices.
.sp
For more details, please see the Tcpreplay Manual at:
http://tcpreplay.appneta.com
.SH "OPTIONS"
.TP
.NOP \f\*[B-Font]\-d\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-dbug\f[]=\f\*[I-Font]number\f[]
Enable debugging output.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 5
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
If configured with \--enable-debug, then you can specify a verbosity 
level for debugging output.  Higher numbers increase verbosity.
.TP
.NOP \f\*[B-Font]\-q\f[], \f\*[B-Font]\-\-quiet\f[]
Quiet mode.
.sp
Print nothing except the statistics at the end of the run
.TP
.NOP \f\*[B-Font]\-T\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-timer\f[]=\f\*[I-Font]string\f[]
Select packet timing mode: select, ioport, gtod, nano.
This option may appear up to 1 times.
The default
\f\*[I-Font]string\f[]
for this option is:
.ti +4
 gtod
.sp
Allows you to select the packet timing method to use:
.sp
.IR "nano"
- Use nanosleep() API
.sp
.IR "select"
- Use select() API
.sp
.IR "ioport"
- Write to the i386 IO Port 0x80
.sp
.IR "gtod [default]"
- Use a gettimeofday() loop
.br
.TP
.NOP \f\*[B-Font]\-\-maxsleep\f[]=\f\*[I-Font]number\f[]
Sleep for no more then X milliseconds between packets.
This option takes an integer number as its argument.
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
Set a limit for the maximum number of milliseconds that tcpreplay will sleep
between packets.  Effectively prevents long delays between packets without
effecting the majority of packets.  Default is disabled.
.TP
.NOP \f\*[B-Font]\-v\f[], \f\*[B-Font]\-\-verbose\f[]
Print decoded packets via tcpdump to STDOUT.
This option may appear up to 1 times.
.sp
.TP
.NOP \f\*[B-Font]\-A\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-decode\f[]=\f\*[I-Font]string\f[]
Arguments passed to tcpdump decoder.
This option may appear up to 1 times.
This option must appear in combination with the following options:
verbose.
.sp
When enabling verbose mode (\fB-v\fP) you may also specify one or more
additional  arguments to pass to \fBtcpdump\fP to modify the way packets
are decoded.  By default, \-n and \-l are used.   Be  sure  to
quote the arguments like: \-A "-axxx" so that they are not interpreted
by tcpreplay.   Please see the tcpdump(1) man page for a complete list of 
options.
.TP
.NOP \f\*[B-Font]\-K\f[], \f\*[B-Font]\-\-preload\-pcap\f[]
Preloads packets into RAM before sending.
.sp
This option loads the specified pcap(s) into RAM before starting to send in order
to improve replay performance while introducing a startup performance hit.
Preloading can be used with or without \fB--loop\fP. This option also suppresses
flow statistics collection for every iteration, which can significantly reduce
memory usage. Flow statistics are predicted based on options supplied and
statistics collected from the first loop iteration.
.TP
.NOP \f\*[B-Font]\-c\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-cachefile\f[]=\f\*[I-Font]string\f[]
Split traffic via a tcpprep cache file.
This option may appear up to 1 times.
This option must appear in combination with the following options:
intf2.
This option must not appear in combination with any of the following options:
dualfile.
.sp
If you have a pcap file you would like to use to send bi-directional
traffic through a device (firewall, router, IDS, etc) then using tcpprep
you can create a cachefile which tcpreplay will use to split the traffic
across two network interfaces.
.TP
.NOP \f\*[B-Font]\-2\f[], \f\*[B-Font]\-\-dualfile\f[]
Replay two files at a time from a network tap.
This option may appear up to 1 times.
This option must appear in combination with the following options:
intf2.
This option must not appear in combination with any of the following options:
cachefile.
.sp
If you captured network traffic using a network tap, then you can end up with
two pcap files- one for each direction.  This option will replay these two
files at the same time, one on each interface and inter-mix them using the 
timestamps in each.
.TP
.NOP \f\*[B-Font]\-i\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-intf1\f[]=\f\*[I-Font]string\f[]
Client to server/RX/primary traffic output interface.
This option may appear up to 1 times.
.sp
Required network interface used to send either all traffic or traffic which is 
marked as 'primary' via tcpprep.  Primary traffic is usually client-to-server 
or inbound (RX) on khial virtual interfaces.
.TP
.NOP \f\*[B-Font]\-I\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-intf2\f[]=\f\*[I-Font]string\f[]
Server to client/TX/secondary traffic output interface.
This option may appear up to 1 times.
.sp
Optional network interface used to send traffic which is marked as 'secondary' 
via tcpprep.  Secondary traffic is usually server-to-client or outbound 
(TX) on khial virtual interfaces.  Generally, it only makes sense to use this
option with \--cachefile.
.TP
.NOP \f\*[B-Font]\-\-listnics\f[]
List available network interfaces and exit.
.sp
.TP
.NOP \f\*[B-Font]\-l\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-loop\f[]=\f\*[I-Font]number\f[]
Loop through the capture file X times.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 0
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 1
.sp
.TP
.NOP \f\*[B-Font]\-\-loopdelay\-ms\f[]=\f\*[I-Font]number\f[]
Delay between loops in milliseconds.
This option must appear in combination with the following options:
loop.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 0
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
.TP
.NOP \f\*[B-Font]\-\-pktlen\f[]
Override the snaplen and use the actual packet len.
This option may appear up to 1 times.
.sp
By default, tcpreplay will send packets based on the size of the "snaplen"
stored in the pcap file which is usually the correct thing to do.  However,
occasionally, tools will store more bytes then told to.  By specifying this
option, tcpreplay will ignore the snaplen field and instead try to send
packets based on the original packet length.  Bad things may happen if
you specify this option.
.TP
.NOP \f\*[B-Font]\-L\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-limit\f[]=\f\*[I-Font]number\f[]
Limit the number of packets to send.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 \-1
.sp
By default, tcpreplay will send all the packets.  Alternatively, you can 
specify a maximum number of packets to send.  
.TP
.NOP \f\*[B-Font]\-\-duration\f[]=\f\*[I-Font]number\f[]
Limit the number of seconds to send.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 \-1
.sp
By default, tcpreplay will send all the packets.  Alternatively, you can
specify a maximum number of seconds to transmit.
.TP
.NOP \f\*[B-Font]\-x\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-multiplier\f[]=\f\*[I-Font]string\f[]
Modify replay speed to a given multiple.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
pps, mbps, oneatatime, topspeed.
.sp
Specify a value to modify the packet replay speed.
Examples:
.nf
        2.0 will replay traffic at twice the speed captured
        0.7 will replay traffic at 70% the speed captured
.fi
.TP
.NOP \f\*[B-Font]\-p\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-pps\f[]=\f\*[I-Font]number\f[]
Replay packets at a given packets/sec.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
multiplier, mbps, oneatatime, topspeed.
This option takes an integer number as its argument.
.sp
.TP
.NOP \f\*[B-Font]\-M\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-mbps\f[]=\f\*[I-Font]string\f[]
Replay packets at a given Mbps.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
multiplier, pps, oneatatime, topspeed.
.sp
Specify a floating point value for the Mbps rate that tcpreplay
should send packets at.
.TP
.NOP \f\*[B-Font]\-t\f[], \f\*[B-Font]\-\-topspeed\f[]
Replay packets as fast as possible.
This option must not appear in combination with any of the following options:
mbps, multiplier, pps, oneatatime.
.sp
.TP
.NOP \f\*[B-Font]\-o\f[], \f\*[B-Font]\-\-oneatatime\f[]
Replay one packet at a time for each user input.
This option must not appear in combination with any of the following options:
mbps, pps, multiplier, topspeed.
.sp
Allows you to step through one or more packets at a time.
.TP
.NOP \f\*[B-Font]\-\-pps\-multi\f[]=\f\*[I-Font]number\f[]
Number of packets to send for each time interval.
This option must appear in combination with the following options:
pps.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 1
.sp
When trying to send packets at very high rates, the time between each packet
can be so short that it is impossible to accurately sleep for the required
period of time.  This option allows you to send multiple packets at a time,
thus allowing for longer sleep times which can be more accurately implemented.
.TP
.NOP \f\*[B-Font]\-\-unique\-ip\f[]
Modify IP addresses each loop iteration to generate unique flows.
.sp
Ensure IPv4 and IPv6 packets will be unique for each \fB--loop\fP iteration. 
This is done in a way that will not alter packet CRC, and therefore will genrally
not affect performance. This option will significantly increase the flows/sec over
generated over multiple loop iterations.
.TP
.NOP \f\*[B-Font]\-\-netmap\f[]
Write packets directly to netmap enabled network adapter.
.sp
This feature will detect netmap capable network drivers on Linux and BSD
systems. If detected, the network driver is bypassed for the execution
duration, and network buffers will be written to directly. This will allow
you to achieve full line rates on commodity network adapters, similar to rates
achieved by commercial network traffic generators. Note that bypassing the network
driver will disrupt other applications connected through the test interface. See
INSTALL for more information.
.sp
This feature can also be enabled by specifying an interface as 'netmap:<intf>'
or 'vale:<intf>. For example 'netmap:eth0' specifies netmap over interface eth0.
.TP
.NOP \f\*[B-Font]\-\-nm\-delay\f[]=\f\*[I-Font]number\f[]
Netmap startup delay.
This option must appear in combination with the following options:
netmap.
This option takes an integer number as its argument.
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 4
.sp
Number of seconds to delay after netmap is loaded. Required to ensure interfaces
are fully up before netmap transmit. Requires netmap option. Default is 4 seconds.
.TP
.NOP \f\*[B-Font]\-\-no\-flow\-stats\f[]
Suppress printing and tracking flow count, rates and expirations.
.sp
Suppress the collection and printing of flow statistics. This option may
improve performance when not using \fB--preload-pcap\fP option, otherwise
its only function is to suppress printing. 
.sp
The flow feature will track and print statistics of the flows being sent. 
A flow is loosely defined as a unique combination of a 5-tuple, i.e. 
source IP, destination IP, source port, destination port and protocol.
.sp
If \fB--loop\fP is specified, the flows from one iteration to the next
will not be unique, unless the packets are altered. Use \fB--unique-ip\fP
or \fBtcpreplay-edit\fP to alter packets between iterations. 
.TP
.NOP \f\*[B-Font]\-\-flow\-expiry\f[]=\f\*[I-Font]number\f[]
Number of inactive seconds before a flow is considered expired.
This option must not appear in combination with any of the following options:
no-flow-stats.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 0
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
This option will track and report flow expirations based on the flow idle
times. The timestamps within the pcap file are used to determine the expiry,
not the actual timestamp of the packets are replayed. For example, a value of
30 suggests that if no traffic is seen on a flow for 30 seconds, any
subsequent traffic would be considered a new flow, and thereby will increment
the flows and flows per second (fps) statistics. 
.sp
This option can be used to optimize flow timeout settings for flow products.
Setting the timeout low may lead to flows being dropped when in fact the flow
is simply slow to respond. Configuring your flow timeouts too high may
increase resources required by your flow product.
.sp
Note that using this option while replaying at higher than original speeds
can lead to inflated flows and fps counts.
.sp
Default is 0 (no expiry) and a typical value is 30-120 seconds.
.TP
.NOP \f\*[B-Font]\-P\f[], \f\*[B-Font]\-\-pid\f[]
Print the PID of tcpreplay at startup.
.sp
.TP
.NOP \f\*[B-Font]\-\-quick\-tx\f[]
Write packets directly to an interface via Quick_TX module.
.sp
This feature will detect detect the presence of the Quick TX module on Linux
systems. If detected, tcpreplay packets will bypass the network stack and
will be written to the network driver via the Quick TX module. This will allow
you to achieve full line rates on commodity network adapters, similar to rates
achieved by commercial network traffic generators. 
.sp
Note that unlike netmap, Quick TX does not bypass the network driver and 
therefore is less disruptive. Yet it can achieve TX rates comparble netmap. 
Quick TX is distributed with Tcpreplay and does not require full kernel source
in order to be installed. See INSTALL for more information.
.sp
This feature can also be enabled by specifying an interface as 'qtx:<intf>'.
For example 'qtx:eth0' specifies Quick TX over interface eth0.
.TP
.NOP \f\*[B-Font]\-\-stats\f[]=\f\*[I-Font]number\f[]
Print statistics every X seconds.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
.sp
Note that this is very much a "best effort" and long delays between
sending packets may cause equally long delays between printing statistics.
.TP
.NOP \f\*[B-Font]\-V\f[], \f\*[B-Font]\-\-version\f[]
Print version information.
.sp
.TP
.NOP \f\*[B-Font]\-h\f[], \f\*[B-Font]\-\-less\-help\f[]
Display less usage information and exit.
.sp
.TP
.NOP \f\*[B-Font]\-H\f[], \f\*[B-Font]\-\-help\f[]
Display usage information and exit.
.TP
.NOP \f\*[B-Font]\-\&!\f[], \f\*[B-Font]\-\-more-help\f[]
Pass the extended usage information through a pager.
.TP
.NOP \f\*[B-Font]\-\-save-opts\f[] [=\f\*[I-Font]cfgfile\f[]]
Save the option state to \fIcfgfile\fP.  The default is the \fIlast\fP
configuration file listed in the \fBOPTION PRESETS\fP section, below.
The command will exit after updating the config file.
.TP
.NOP \f\*[B-Font]\-\-load-opts\f[]=\f\*[I-Font]cfgfile\f[], \f\*[B-Font]\-\-no-load-opts\f[]
Load options from \fIcfgfile\fP.
The \fIno-load-opts\fP form will disable the loading
of earlier config/rc/ini files.  \fI\-\-no-load-opts\fP is handled early,
out of order.
.PP
.SH "OPTION PRESETS"
Any option that is not marked as \fInot presettable\fP may be preset
by loading values from configuration ("RC" or ".INI") file(s).
The \fIhomerc\fP file is "\fI$$/\fP", unless that is a directory.
In that case, the file "\fI.tcpreplayrc\fP"
is searched for within that directory.
.SH "FILES"
See \fBOPTION PRESETS\fP for configuration files.
.SH "EXIT STATUS"
One of the following exit values will be returned:
.TP
.NOP 0 " (EXIT_SUCCESS)"
Successful program execution.
.TP
.NOP 1 " (EXIT_FAILURE)"
The operation failed or the command syntax was not valid.
.TP
.NOP 66 " (EX_NOINPUT)"
A specified configuration file could not be loaded.
.TP
.NOP 70 " (EX_SOFTWARE)"
libopts had an internal operational error.  Please report
it to autogen-users@lists.sourceforge.net.  Thank you.
.PP
.SH "AUTHORS"
Copyright 2013-2014 Fred Klassen \- AppNeta
Copyright 2000-2012 Aaron Turner
For support please use the tcpreplay-users@lists.sourceforge.net mailing list.
The latest version of this software is always available from:
http://tcpreplay.appneta.com/
.SH "COPYRIGHT"
Copyright (C) 2000-2014 Aaron Turner and Fred Klassen all rights reserved.
This program is released under the terms of the GNU General Public License, version 3 or later.
.SH "BUGS"
Please send bug reports to: tcpreplay-users@lists.sourceforge.net
.SH "NOTES"
This manual page was \fIAutoGen\fP-erated from the \fBtcpreplay\fP
option definitions.


use-cases/elastic_router/ddtcpreplay/src/tcpreplay.c
/* $Id$ */

/*
 *   Copyright (c) 2001-2010 Aaron Turner <aturner at synfin dot net>
 *   Copyright (c) 2013-2014 Fred Klassen <tcpreplay at appneta dot com> -
 *AppNeta
 *
 *   The Tcpreplay Suite of tools is free software: you can redistribute it
 *   and/or modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation, either version 3 of the
 *   License, or with the authors permission any later version.
 *
 *   The Tcpreplay Suite is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with the Tcpreplay Suite.  If not, see
 *<http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include "defines.h"
#include "common.h"

#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include <pthread.h>

#include "tcpreplay.h"
#include "tcpreplay_api.h"

#ifdef TCPREPLAY_EDIT
#include "tcpreplay_edit_opts.h"
#include "tcpedit/tcpedit.h"
tcpedit_t *tcpedit;
#else
#include "tcpreplay_opts.h"
#endif

#include "send_packets.h"
#include "replay.h"
#include "signal_handler.h"
#include "dd.h"
#ifdef DEBUG
int debug = 0;
#endif

int run_loop = 0;
pthread_mutex_t dd_startm = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t dd_runc = PTHREAD_COND_INITIALIZER;

tcpreplay_t *ctx;

#define ST_RISE 0
#define ST_HIGH 1
#define ST_FALL 2
#define ST_LOW 3
// Start at the low state
char loop_state = ST_LOW;

void flow_stats(const tcpreplay_t *ctx, bool unique_ip);
// callback functions
void on_reg(void *args) {
  dd_t *dd = (dd_t *)args;
  printf("\nRegistered with broker %s!\n", dd_get_endpoint(dd));
  dd_subscribe(dd, "tcpreplay", "all");
  fflush(stdout);
}

void on_discon(void *args) {
  dd_t *dd = (dd_t *)args;
  printf("\nGot disconnected from broker %s!\n", dd_get_endpoint(dd));
  fflush(stdout);
}

int handle_data(unsigned char *data) {
  if (streq("low", data)) {
    if (loop_state == ST_HIGH) {
      printf("Transitioning to a lower flow\n");
      loop_state = ST_FALL;
      return 0;
    } else {
      printf("Got request to go low when not high, ignoring!\n");
      return 0;
    }
  } else if (streq("high", data)) {
    if (loop_state == ST_LOW) {
      printf("Transitioning to a higher flow\n");
      loop_state = ST_RISE;
      return 0;
    } else {
      printf("Got request to go high when not low, ignoring!\n");
      return 0;
    }
  } else if (streq("start", data)) {
    printf("starting!\n");
    pthread_mutex_lock(&dd_startm);
    run_loop = 1;
    pthread_cond_signal(&dd_runc);
    pthread_mutex_unlock(&dd_startm);
  } else if (streq("stop", data)) {
    printf("stopping!\n");
    pthread_mutex_lock(&dd_startm);
    run_loop = 0;
    pthread_cond_signal(&dd_runc);
    pthread_mutex_unlock(&dd_startm);
  }
  return -1;
}
void on_pub(char *source, char *topic, unsigned char *data, int length,
            void *args) {
  dd_t *dd = (dd_t *)args;
  int rc = handle_data(data);
  if (rc == -1)
    printf("\nPUB S: %s T: %s L: %d D: '%s'", source, topic, length, data);
  fflush(stdout);
}

void on_data(char *source, unsigned char *data, int length, void *args) {
  dd_t *dd = (dd_t *)args;
  int rc = handle_data(data);
  if (rc == -1)
    printf("\nDATA S: %s L: %d D: '%s'", source, length, data);
  fflush(stdout);
}
void on_error(int error_code, char *error_message, void *args) {
  switch (error_code) {
  case DD_ERROR_NODST:
    printf("Error - no destination: %s\n", error_message);
    break;
  case DD_ERROR_REGFAIL:
    printf("Error - registration failed: %s\n", error_message);
    break;
  case DD_ERROR_VERSION:
    printf("Error - version: %s\n", error_message);
    break;
  default:
    printf("Error - unknown error!\n");
    break;
  }
  fflush(stdout);
}

int main(int argc, char *argv[]) {
  int i, optct = 0;
  int rcode;
  char buf[1024];

  fflush(NULL);

  ctx = tcpreplay_init();
#ifdef TCPREPLAY
  optct = optionProcess(&tcpreplayOptions, argc, argv);
#elif defined TCPREPLAY_EDIT
  optct = optionProcess(&tcpreplay_editOptions, argc, argv);
#endif
  argc -= optct;
  argv += optct;

  fflush(NULL);
  rcode = tcpreplay_post_args(ctx, argc);
  if (rcode <= -2) {
    warnx("%s", tcpreplay_getwarn(ctx));
  } else if (rcode == -1) {
    errx(-1, "Unable to parse args: %s", tcpreplay_geterr(ctx));
  }

  fflush(NULL);
#ifdef TCPREPLAY_EDIT
  /* init tcpedit context */
  if (tcpedit_init(&tcpedit, sendpacket_get_dlt(ctx->intf1)) < 0) {
    errx(-1, "Error initializing tcpedit: %s", tcpedit_geterr(tcpedit));
  }

  /* parse the tcpedit args */
  rcode = tcpedit_post_args(tcpedit);
  if (rcode < 0) {
    errx(-1, "Unable to parse args: %s", tcpedit_geterr(tcpedit));
  } else if (rcode == 1) {
    warnx("%s", tcpedit_geterr(tcpedit));
  }

  if (tcpedit_validate(tcpedit) < 0) {
    errx(-1, "Unable to edit packets given options:\n%s",
         tcpedit_geterr(tcpedit));
  }
#endif

  if (ctx->options->preload_pcap && !HAVE_OPT(QUIET)) {
    notice("File Cache is enabled");
  }

  /*
   * Setup up the file cache, if required
   */
  if (ctx->options->preload_pcap) {
    /* Initialise each of the file cache structures */
    for (i = 0; i < argc; i++) {
      ctx->options->file_cache[i].index = i;
      ctx->options->file_cache[i].cached = FALSE;
      ctx->options->file_cache[i].packet_cache = NULL;
    }
  }

  for (i = 0; i < argc; i++) {
    tcpreplay_add_pcapfile(ctx, argv[i]);

    /* preload our pcap file? */
    if (ctx->options->preload_pcap) {
      preload_pcap_file(ctx, i);
    }
  }
  if (i < 4) {
    notice("\nFailed: ddtcpreplay needs four pcap files to play!\n"
           "E.g. tcpreplay -K -i dummy0 -l 0 rise-er1.pcap high-er1.pcap "
           "fall-er1.pcap low-er1.pcap\n");
    exit(-1);
  }
  char *ddname = getenv("DDNAME");
  if (ddname == NULL) {
    ddname = "tcpreplay";
    printf("env variable DDNAME empty, using %s\n", ddname);
  }
  char *dealer = getenv("DEALER");
  if (dealer == NULL) {
    dealer = "tcp://127.0.0.1:5555";
    printf("env variable DEALER empty, using %s\n", dealer);
  }
  char *keyfile = getenv("KEYFILE");
  if (keyfile == NULL) {
    keyfile = "/etc/doubledecker/public-keys.json";
    printf("env variable KEYFILE empty, using %s\n", keyfile);
  }
  ctx->ddclient = dd_new(ddname, dealer, keyfile, on_reg, on_discon, on_data, on_pub,
                    on_error);

  /* init the signal handlers */
  init_signal_handlers();

  /* main loop */
  rcode = tcpreplay_replay(ctx);

  
  if (rcode < 0) {
    notice("\nFailed: %s\n", tcpreplay_geterr(ctx));
    exit(rcode);
  } else if (rcode == 1) {
    notice("\nWarning: %s\n", tcpreplay_getwarn(ctx));
  }

  if (ctx->stats.bytes_sent > 0) {
    packet_stats(&ctx->stats);
    if (ctx->options->flow_stats)
      flow_stats(ctx, ctx->options->unique_ip
#ifdef TCPREPLAY_EDIT
                          || tcpedit->seed
#endif
                 );
    sendpacket_getstat(ctx->intf1, buf, sizeof(buf));
    printf("%s", buf);
    if (ctx->intf2 != NULL) {
      sendpacket_getstat(ctx->intf2, buf, sizeof(buf));
      printf("%s", buf);
    }
  }
  tcpreplay_close(ctx);
  dd_destroy(&ctx->ddclient);
  return 0;
} /* main() */

/**
 * Print various flow statistics
 */
void flow_stats(const tcpreplay_t *ctx, bool unique_ip) {
  struct timeval diff;
  COUNTER diff_us;
  const tcpreplay_stats_t *stats = &ctx->stats;
  COUNTER flows_total = stats->flows;
  COUNTER flows_unique = stats->flows_unique;
  COUNTER flows_expired = stats->flows_expired;
  COUNTER flow_packets;
  COUNTER flow_non_flow_packets;
  COUNTER flows_sec = 0;
  u_int32_t flows_sec_100ths = 0;

  timersub(&stats->end_time, &stats->start_time, &diff);
  diff_us = TIMEVAL_TO_MICROSEC(&diff);

  if (!flows_total || !ctx->iteration)
    return;

  /*
   * When packets are read into cache,  flows
   * are only counted in first iteration
   * If flows are unique from one loop iteration
   * to the next then multiply by the number of
   * successful iterations.
   */
  if (unique_ip && ctx->options->preload_pcap) {
    flows_total *= ctx->iteration;
    flows_unique *= ctx->iteration;
    flows_expired *= ctx->iteration;
  }

  flow_packets = stats->flow_packets * ctx->iteration;
  flow_non_flow_packets = stats->flow_non_flow_packets * ctx->iteration;

  if (diff_us) {
    COUNTER flows_sec_X100;

    flows_sec_X100 = (flows_total * 100 * 1000 * 1000) / diff_us;
    flows_sec = flows_sec_X100 / 100;
    flows_sec_100ths = flows_sec_X100 % 100;
  }

  if (ctx->options->flow_expiry)
    printf("Flows: " COUNTER_SPEC " flows, " COUNTER_SPEC
           " unique, " COUNTER_SPEC " expired, %llu.%02u fps, " COUNTER_SPEC
           " flow packets, " COUNTER_SPEC " non-flow\n",
           flows_total, flows_unique, flows_expired, flows_sec,
           flows_sec_100ths, flow_packets, flow_non_flow_packets);
  else
    printf("Flows: " COUNTER_SPEC " flows, %llu.%02u fps, " COUNTER_SPEC
           " flow packets, " COUNTER_SPEC " non-flow\n",
           flows_total, flows_sec, flows_sec_100ths, flow_packets,
           flow_non_flow_packets);
}

/* vim: set tabstop=8 expandtab shiftwidth=4 softtabstop=4: */
